app/src/main/java/cloudcity/GPSMonitor.java

```diff
@@ -62,10 +86,83 @@ public void startMonitoring() {
         List<String> allGpsProviders = locationManager.getAllProviders();
         Log.d(TAG, "all GPS providers: "+allGpsProviders);
         locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, GPS_POLLING_SPEED_IN_MS, GPS_POLLING_MIN_DIST_IN_M, locationListener);
+
+        Log.d(TAG, "Starting speed polling task...");
+
+        valueMonitor = new ValueMonitor();
+        valueMonitor.setCallback(() -> {
+            Iperf3Monitor.getInstance().startDefault15secTest();
+        });
+        valueMonitor.startMonitoring();
+
         Log.d(TAG, "<-- startMonitoring()");
     }
 
     public void stopMonitoring() {
         locationManager.removeUpdates(locationListener);
     }
+
+    private class ValueMonitor {
+        private static final String TAG = "ValueMonitor";
+
+        private final TimerWrapper timer;
+        /**
+         * Callback that will be invoked when the monitored value has been under {@link #THRESHOLD_VALUE}
+         * for at least {@link #THRESHOLD_DURATION} millis
+         */
+        private @Nullable ValueMonitorCallback callback;
+
+        public ValueMonitor() {
+            // Start the timer as a daemon thread
+            timer = new TimerWrapper(true);
+        }
+
+        /**
+         * Set or clear the {@link #callback} parameter
+         * @param callback the new callback to set
+         */
+        public void setCallback(@Nullable ValueMonitorCallback callback) {
+            this.callback = callback;
+        }
+
+        public void startMonitoring() {
+            timer.schedule(new TimerTask() {
+                @Override
+                public void run() {
+                    monitorValue();
+                }
+            }, 0, VALUE_MONITOR_INTERVAL);
+        }
+
+        public void stopMonitoring() {
+            timer.stop();
+        }
+
+        /**
+         * Logic to monitor the value and track the time under threshold
+         */
+        private void monitorValue() {
+            if (lastSpeed < THRESHOLD_VALUE) {
+                // If value is under threshold, increment the time under threshold
+                timeUnderThreshold += VALUE_MONITOR_INTERVAL;
+
+                // Check if the time under threshold exceeds 5 seconds
+                if (timeUnderThreshold >= THRESHOLD_DURATION) {
+                    Log.d(TAG, "value has been under threshold for "+timeUnderThreshold+"ms, firing callback");
+                    // Trigger the callback if the condition is met
+                    callback.onUnderThresholdValueForAtLeastThresholdDuration();
+
+                    // Reset the time tracking after triggering callback
+                    timeUnderThreshold = 0;
+                }
+            } else {
+                // If the value is above the threshold, reset the time tracking
+                timeUnderThreshold = 0;
+            }
```

Comment:
no, it's ok. we want the callback to trigger for every 5 seconds that the speed has been under threshold, and then the callbacker (callback object) will make a call to Iperf3Monitor to run a new iperf3 test which lasts 15 secs, so that at least 2 callbacks will be discarded because a test is already running. more likely 3 will be discarded.

there's no need for such a boolean. resolving.